### PR TITLE
New kyma installation

### PR DESCRIPTION
**Pull request by jasiu001**
Changes proposed in this pull request:

- New installation process for provisioner component based on parallel installation

### CODE REVIEW

File: components/provisioner/pkg/gqlschema/schema.graphql [hunk 1]
```diff
@@ -215,12 +215,18 @@ input OpenStackProviderConfigInput {
...
 
+enum KymaInstallationMethod {
+    KymaOperator
+    ParallelInstall
+}
+
 input KymaConfigInput {
     version: String!                            # Kyma version to install on the cluster
     profile: KymaProfile                        # Optional resources profile
+    kymaInstaller: KymaInstallationMethod       # Optional installation method (KymaOperator is default)
```
Comment (akgalwas) [inline components/provisioner/pkg/gqlschema/schema.graphql:226]: We should also add kymaInstaller field to KymaConfig type to be able to return it in `runtimeStatus` query.

File: components/provisioner/pkg/gqlschema/schema.graphql [hunk 2]
```diff
@@ -232,9 +238,10 @@ input ConfigEntryInput {
 input ComponentConfigurationInput {
     component: String!                    # Kyma component name
     namespace: String!                    # Namespace to which component should be installed
+    prerequisite: Boolean                 # Specifies if component is prerequisite or it is optional
```
Comment (akgalwas) [inline components/provisioner/pkg/gqlschema/schema.graphql:241]: We should also add prerequisite field to KymaConfig type to be able to return it in `runtimeStatus` query.

File: components/provisioner/cmd/main.go [hunk 3]
```diff
@@ -177,6 +181,26 @@ func main() {
 	dbsFactory := dbsession.NewFactory(connection)
 	installationService := installation.NewInstallationService(cfg.ProvisioningTimeout.Installation, installationHandlerConstructor, cfg.Gardener.ClusterCleanupResourceSelector)
 
+	downloadManager := download.NewManager(&sync.Mutex{}, download.Config{
+		KymaURL:              "https://github.com/kyma-project/kyma",
+		ResourcesPathTmp:     "/app/downloads/%s",
```
Comment (akgalwas) [inline components/provisioner/cmd/main.go:186]: It doesn't work as there is not volume with write permissions mounted to the pod. Provisioner's deployment needs to be modified.

**Review by akgalwas — CHANGES_REQUESTED** [review]
This is a first batch of comments.

File: components/provisioner/internal/operations/stages/provisioning/wait_for_installation.go [hunk 4]
```diff
@@ -50,7 +55,12 @@ func (s *WaitForInstallationStep) Run(cluster model.Cluster, operation model.Ope
 		return operations.StageResult{}, fmt.Errorf("error: failed to create kubernetes config from raw: %s", err.Error())
 	}
 
-	installationState, err := s.installationClient.CheckInstallationState(k8sConfig)
+	client, ok := s.installationClients[cluster.KymaConfig.Installer]
+	if !ok {
+		return operations.StageResult{}, fmt.Errorf("error: installation client for installation %s does not exist", cluster.KymaConfig.Installer)
+	}
+
+	installationState, err := client.CheckInstallationState(cluster.ID, k8sConfig)
```
Comment (akgalwas) [inline components/provisioner/internal/operations/stages/provisioning/wait_for_installation.go:63]: When installation reaches the timeout -it is reported in logs, installation operation is recorded in DB in the failed state with `error: timeout while processing operation`. But installer workers are not stopped and the installation process is not interrupted on the cluster.

File: components/provisioner/internal/installation/download/components.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
+package download
```
Comment (akgalwas) [inline components/provisioner/internal/installation/download/components.go:1]: I would consider changing package structure a bit. I would suggest to group components related to parallel-install and Kyma Operator in a separate folder structure.

File: components/provisioner/internal/installation/parallel_installation.go [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+		case deployment.ProcessStart:
+			p.log.Infof("Starting installation phase '%s'", update.Phase)
+			showCompStatus(update.Component)
+		case deployment.ProcessRunning:
+			showCompStatus(update.Component)
+			consumeEvent(update)
+		case deployment.ProcessFinished:
+			p.log.Infof("Finished installation phase '%s' successfully", update.Phase)
+		case deployment.ProcessExecutionFailure, deployment.ProcessForceQuitFailure, deployment.ProcessTimeoutFailure:
+			p.log.Errorf("Installation failed on component %s, status: %s, error: %v", update.Component.Name, update.Event, update.Error)
```
[the remark this answers is not in the document]
Comment (jasiu001) [reply components/provisioner/internal/installation/parallel_installation.go:98]: for logger using by `deployer` (passing as an argument in config), I added runtimeID value as a tag

File: components/provisioner/internal/installation/components_status.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+	cmpList := make(map[string]bool)
+	for _, component := range components {
+		cmpKey := componentKey(string(component.Component), component.Namespace)
+		cmpList[cmpKey] = false
+	}
+
+	return &ComponentsStatus{components: cmpList}
+}
+
+func (cs *ComponentsStatus) ConsumeEvent(event deployment.ProcessUpdate) {
```
Comment (ralikio) [inline components/provisioner/internal/installation/components_status.go:28]: During testing when timeout occured for application-connector/xip-patch components the invocation returned nil error with main truncated trace of:

components_status.go:33 
parallel_installation.go:83
parallel_installation.go:99

File: components/provisioner/internal/installation/download/components/tar.go [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+	"github.com/pkg/errors"
+)
+
+func DownloadTgz(URL string, tmpDst string) error {
+	path, err := download.GetFile(URL, tmpDst)
+	if err != nil {
+		return errors.Wrap(err, "while downloading tgz archive")
+	}
+
+	err = archive.Untar(path, tmpDst)
```
Comment (ralikio) [inline components/provisioner/internal/installation/download/components/tar.go:18]: Utility does not support all tar extensions. I was unable to install component using tar.gz extension (tgz works as expected).

File: components/provisioner/internal/installation/download/components/tar.go [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+	if err != nil {
+		return errors.Wrap(err, "while unzipping tgz archive")
```
Comment (ralikio) [inline components/provisioner/internal/installation/download/components/tar.go:20]: Logic does not support "file://" protocols.

File: components/provisioner/internal/parallel-installation/components_status.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+		cmpKey := componentKey(string(component.Component), component.Namespace)
+		cmpList[cmpKey] = false
+	}
+
+	return &ComponentsStatus{components: cmpList}
+}
+
+func (cs *ComponentsStatus) ConsumeEvent(event deployment.ProcessUpdate) {
+	// check if installation process failed
+	if event.Event == deployment.ProcessExecutionFailure ||
```
Comment (akgalwas) [inline components/provisioner/internal/parallel-installation/components_status.go:30]: During tests I came across the following crash:
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation cod
e=0x1 addr=0x0 pc=0x1d60bd2]

goroutine 6737 [running]:
github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.(*ComponentsStatus).ConsumeEvent(0x0, 0x239565b, 0xe, 0x239ac17, 0x11, 0x0, 0x0, 0xc004a17110, 0x15, 0xc000b9ba00, ...)
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/components_status.go:38 +0x152
github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.parallelInstallationService.getCallbackUpdate.func2(0x239565b, 0xe, 0x239ac17, 0x11, 0x0, 0x0, 0xc004a17110, 0x15, 0xc000b9ba00, 0x10, ...)
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/parallel_installation.go:84 +0xbc
github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.parallelInstallationService.getCallbackUpdate.func3(0x239565b, 0xe, 0x239ac17, 0x11, 0x0, 0x0, 0xc004a17110, 0x15, 0xc000b9ba00, 0x10, ...)
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/parallel_installation.go:95 +0x335
github.com/kyma-incubator/hydroform/parallel-install/pkg/deployment.(*core).processUpdateComponent(0xc002690810, 0x239ac17, 0x11, 0xc004a17110, 0x15, 0xc000b9ba00, 0x10, 0xc000b9b100, 0xa, 0x238ceea, ...)
	/go/pkg/mod/github.com/kyma-incubator/hydroform/parallel-install@v0.0.0-20210610074205-bbaf6958f1a5/pkg/deployment/core.go:115 +0x155


github.com/kyma-incubator/hydroform/parallel-install/pkg/deployment.(*Deployment).deployComponents(0xc0022681a8, 0x26af3c0, 0xc0071455c0, 0xc007354840, 0x239ac17, 0x11, 0xc007145580, 0x1174f2ca8ea, 0x15d289160ea, 0x0, ...)
	/go/pkg/mod/github.com/kyma-incubator/hydroform/parallel-install@v0.0.0-20210610074205-bbaf6958f1a5/pkg/deployment/deployment.go:153 +0x358
github.com/kyma-incubator/hydroform/parallel-install/pkg/deployment.(*Deployment).startKymaDeployment(0xc0022681a8, 0x2677b78, 0xc007145500, 0xc007145540, 0xc007145580, 0x0, 0x0)
	/go/pkg/mod/github.com/kyma-incubator/hydroform/parallel-install@v0.0.0-20210610074205-bbaf6958f1a5/pkg/deployment/deployment.go:130 +0x5d8
github.com/kyma-incubator/hydroform/parallel-install/pkg/deployment.(*Deployment).StartKymaDeployment(0xc0022681a8, 0xc006652240, 0xc001987a90)
	/go/pkg/mod/github.com/kyma-incubator/hydroform/parallel-install@v0.0.0-20210610074205-bbaf6958f1a5/pkg/deployment/deployment.go:85 +0x2d6
github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.(*Deployer).StartKymaDeployment(0xc00263a1e0, 0x0, 0x0)
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/deployer.go:24 +0x7a
github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.AsyncDeployment.StartKymaDeployment.func1(0x2661c60, 0xc00263a1e0, 0x2494ff8, 0x2495000)
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/parallel_installation.go:40 +0x31
created by github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.AsyncDeployment.StartKymaDeployment
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/parallel_installation.go:39 +0x5d
```

I suppose that this behaviour is caused by the fact, that CheckInstallationStatus removes entry for RuntimeID when the installation fails or succeeds. If there is an event sent by parallel-install after installation completed event object will be nil. We can add nil checks to solve the issue, but parallel-install shouldn't sent any update after installation finished.

File: components/provisioner/internal/parallel-installation/components_status.go [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+		event.Event == deployment.ProcessForceQuitFailure ||
+		event.Event == deployment.ProcessTimeoutFailure {
+		cs.installationFailed = fmt.Errorf("status: %s, error: %v", event.Event, event.Error)
+	}
+
+	// check if component installation process succeeded
+	key := componentKey(event.Component.Name, event.Component.Namespace)
+	if _, ok := cs.components[key]; !ok {
```
Comment (akgalwas) [inline components/provisioner/internal/parallel-installation/components_status.go:38]: Another crash observed during stress tests:
```
goroutine 6737 [running]:
github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.(*ComponentsStatus).ConsumeEvent(0x0, 0x239565b, 0xe, 0x239ac17, 0x11, 0x0, 0x0, 0xc004a17110, 0x15, 0xc000b9ba00, ...)
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/components_status.go:38 +0x152
github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.parallelInstallationService.getCallbackUpdate.func2(0x239565b, 0xe, 0x239ac17, 0x11, 0x0, 0x0, 0xc004a17110, 0x15, 0xc000b9ba00, 0x10, ...)
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/parallel_installation.go:84 +0xbc
github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.parallelInstallationService.getCallbackUpdate.func3(0x239565b, 0xe, 0x239ac17, 0x11, 0x0, 0x0, 0xc004a17110, 0x15, 0xc000b9ba00, 0x10, ...)
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/parallel_installation.go:95 +0x335
github.com/kyma-incubator/hydroform/parallel-install/pkg/deployment.(*core).processUpdateComponent(0xc002690810, 0x239ac17, 0x11, 0xc004a17110, 0x15, 0xc000b9ba00, 0x10, 0xc000b9b100, 0xa, 0x238ceea, ...)
	/go/pkg/mod/github.com/kyma-incubator/hydroform/parallel-install@v0.0.0-20210610074205-bbaf6958f1a5/pkg/deployment/core.go:115 +0x155
github.com/kyma-incubator/hydroform/parallel-install/pkg/deployment.(*Deployment).deployComponents(0xc0022681a8, 0x26af3c0, 0xc0071455c0, 0xc007354840, 0x239ac17, 0x11, 0xc007145580, 0x1174f2ca8ea, 0x15d289160ea, 0x0, ...)
	/go/pkg/mod/github.com/kyma-incubator/hydroform/parallel-install@v0.0.0-20210610074205-bbaf6958f1a5/pkg/deployment/deployment.go:153 +0x358
github.com/kyma-incubator/hydroform/parallel-install/pkg/deployment.(*Deployment).startKymaDeployment(0xc0022681a8, 0x2677b78, 0xc007145500, 0xc007145540, 0xc007145580, 0x0, 0x0)
	/go/pkg/mod/github.com/kyma-incubator/hydroform/parallel-install@v0.0.0-20210610074205-bbaf6958f1a5/pkg/deployment/deployment.go:130 +0x5d8
github.com/kyma-incubator/hydroform/parallel-install/pkg/deployment.(*Deployment).StartKymaDeployment(0xc0022681a8, 0xc006652240, 0xc001987a90)
	/go/pkg/mod/github.com/kyma-incubator/hydroform/parallel-install@v0.0.0-20210610074205-bbaf6958f1a5/pkg/deployment/deployment.go:85 +0x2d6
github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.(*Deployer).StartKymaDeployment(0xc00263a1e0, 0x0, 0x0)
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/deployer.go:24 +0x7a
github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.AsyncDeployment.StartKymaDeployment.func1(0x2661c60, 0xc00263a1e0, 0x2494ff8, 0x2495000)
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/parallel_installation.go:40 +0x31
created by github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation.AsyncDeployment.StartKymaDeployment
	/go/src/github.com/kyma-project/control-plane/components/provisioner/internal/parallel-installation/parallel_installation.go:39 +0x5d
```

At the first glance I cannot tell what was the cause here.

File: components/provisioner/internal/installation/parallel_installation.go [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+	}
+
+	return installation.InstallationState{State: installation.NoInstallationState}, nil
+}
+
+func (p parallelInstallationService) TriggerInstallation(runtimeID, kubeconfigRaw string, kymaProfile *model.KymaProfile, release model.Release, globalConfig model.Configuration, componentsConfig []model.KymaComponentConfig) error {
+	p.log.Info("Installation triggered")
+
+	// download all resources
+	p.log.Info("Collect all require components")
```
[the remark this answers is not in the document]
Comment (jasiu001) [reply components/provisioner/internal/installation/parallel_installation.go]: A new step which is responsible for downloading all necessary resources was added. The step is run first so all errors related to the wrong sourceURL will appear immediately.

File: components/provisioner/internal/installation/parallel_installation.go [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+		InstallationResourcePath:      installationResourcePath,
+		KubeconfigSource: config.KubeconfigSource{
+			Content: kubeconfigRaw,
+		},
+		Version: release.Version,
+	}
+
+	// create Kyma deployer and start deployment
+	p.log.Info("Start deployment process")
+	deployer := p.createDeployer(runtimeID, installationCfg, builder, p.getCallbackUpdate)
```
[the remark this answers is not in the document]
Comment (jasiu001) [reply components/provisioner/internal/installation/parallel_installation.go]: I made a test with an incorrect component name: I got the following error from the provisioner:
```
Installation failed on component bad-component, status: ProcessExecutionFailure, error: <nil>
Installation failed on component , status: ProcessExecutionFailure, error: Kyma deployment failed due to errors in 1 component(s)
(...)
Installation failed: while starting Kyma deployment: Kyma deployment failed due to errors in 1 component(s)
```
I see two problems here related to `deployment.ProcessUpdate` event from the installer:
- installer sent two events one for a component without an error message and one without a component name and with error message
- error message not saying much about failing (both error from event and error from `StartKymaDeployment()` method)

Additionally, there are logs directly from the installer which says more:
```
Error deploying bad-component: Error: Failed to deploy bad-component within the configured time. Error: stat /app/downloads/b3cc43dd-49c5-4f06-a9bc-d662989997bf/resources/bad-component: no such file or directory
(...)
Component: bad-component, Status: Error
```

IMO events and error message from the installer should be fixed, from the provisioner perspective logs are fine, what do you think? Do you expect more information from the provisioner?

File: components/provisioner/internal/installation/overrides.go [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+
+	"github.com/pkg/errors"
+)
+
+//go:generate mockery -name=OverrideBuilder
+type OverrideBuilder interface {
+	AddOverrides(string, map[string]interface{}) error
+}
+
+func SetOverrides(ob OverrideBuilder, components []model.KymaComponentConfig, globalConfiguration model.Configuration) error {
```
[the remark this answers is not in the document]
Comment (jasiu001) [reply components/provisioner/internal/installation/overrides.go]: new validation for keys is added, now error message tells more:
```
time="2021-06-18T12:45:38Z" level=error msg="error while processing operation, stage failed: error: failed to start installation: while set overrides to the OverridesBuilder: key for override \".Global.some.elo=yolo\" is incorrect: no key segment can be empty" Component=Executor OperationId=9532e440-db62-4410-a066-5843376b9fcc OperationType=PROVISION RuntimeId=775c7b49-e9f4-4ccb-b1e7-ffb331cc63c9 ShootName=c-52a0b87 Stage=StartingInstallation
```

File: components/provisioner/internal/operations/stages/provisioning/wait_for_installation.go (hunk 4 above)
Comment (jasiu001) [reply components/provisioner/internal/operations/stages/provisioning/wait_for_installation.go:63]: The installer and its method `StartKymaDeployment()` doesn't support any channel which could interrupt the installation process and stop goroutine with installation execution. However, it supports [timeout parameter](https://github.com/kyma-project/hydroform/blob/main/parallel-install/example/example.go#L67) which [stops the installation process](https://github.com/kyma-project/hydroform/blob/main/parallel-install/pkg/deployment/deployment.go#L182) based on the value of this parameter. The solution, in this case, could be set `CancelTimeout`  parameter the same as timeout for `wait_for_installlation`. What do you think about it?

File: components/provisioner/internal/parallel-installation/components_status.go (hunk 10 above)
Comment (jasiu001) [reply components/provisioner/internal/parallel-installation/components_status.go:30]: I updated `consumeEvent` function inside the callback method:
```go
consumeEvent := func(event deployment.ProcessUpdate) {
   p.mux.Lock()
   if _, ok := p.installationStatus[runtimeID]; ok {
      p.installationStatus[runtimeID].ConsumeEvent(event)
   } else {
      p.log.Warnf("status checker for runtime: %s not exist anymore", runtimeID)
   }
   p.mux.Unlock()
}
```
when the installation status checker with specific runtimeID will be removed then `ConsumeEvent` method won't be triggered.
I also add immidietally return from `ConsumeEvent` if event type suggests installation error.

File: components/provisioner/internal/parallel-installation/components_status.go (hunk 11 above)
Comment (jasiu001) [reply components/provisioner/internal/parallel-installation/components_status.go:39]: It looks similar to [previous](https://github.com/kyma-project/control-plane/pull/678#discussion_r653338558) issue (lines are the same).

File: components/provisioner/internal/operations/stages/provisioning/wait_for_installation.go (hunk 4 above)
Comment (jasiu001) [reply components/provisioner/internal/operations/stages/provisioning/wait_for_installation.go:63]: After discussion, I added a timeout to the parallel installer equal to timeout for `wait_for_installlation` step.

File: components/provisioner/internal/installation/download/components/tar.go (hunk 8 above)
Comment (jasiu001) [reply components/provisioner/internal/installation/download/components/tar.go]: After discussion, it looks like an extra feature for future implementation.

File: components/provisioner/internal/installation/download/components/tar.go (hunk 9 above)
Comment (jasiu001) [reply components/provisioner/internal/installation/download/components/tar.go]: After discussion, it looks like an extra feature for future implementation.

File: components/provisioner/internal/installation/components_status.go (hunk 7 above)
Comment (jasiu001) [reply components/provisioner/internal/installation/components_status.go]: It looks similar to [this issue](https://github.com/kyma-project/control-plane/pull/678#discussion_r654344751). Installer returns event without error inside, e.g.:
```
Installation failed on component bad-component, status: ProcessExecutionFailure, error: <nil>
```
will be reported to the parallel installer team

File: components/provisioner/internal/operations/stages/provisioning/wait_for_installation.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jasiu001) [reply components/provisioner/internal/operations/stages/provisioning/wait_for_installation.go:63]: Delegated to the parallel install team: [#317](https://github.com/kyma-project/hydroform/issues/317)

File: components/provisioner/internal/installation/parallel_installation.go [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+	p.installationStatus[runtimeID] = NewComponentsStatus(componentsConfig)
+
+	asyncDeployment := &AsyncDeployment{deployer}
+	asyncDeployment.StartKymaDeployment(callbackSuccess, callbackErrors)
```
[the remark this answers is not in the document]
Comment (jasiu001) [reply components/provisioner/internal/installation/parallel_installation.go]: Delegated to the parallel install team: [#317](https://github.com/kyma-project/hydroform/issues/317)

File: components/provisioner/internal/installation/parallel_installation.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (jasiu001) [reply components/provisioner/internal/installation/parallel_installation.go]: Delegated to the parallel install team: [#320](https://github.com/kyma-project/hydroform/issues/320)

File: components/provisioner/internal/installation/parallel_installation.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (jasiu001) [reply components/provisioner/internal/installation/parallel_installation.go]: After discussion: downloading all necessary resources were moved to the `ProvisionRuntime()` method (`/internal/provisioning/service.go`) before triggering cluster creation.

**Comment by jasiu001** [comment]
Deprecated.

**Pull request closed without merging**